crates/sc-consensus-subspace-rpc/src/lib.rs

```diff
@@ -658,21 +657,23 @@ where
             .collect())
     }
 
-    fn piece(&self, piece_index: PieceIndex) -> RpcResult<Option<Vec<u8>>> {
-        if let Some(piece_provider) = self.piece_provider.as_ref() {
-            let result = piece_provider.get_piece_by_index(piece_index).map_err(|_| {
-                JsonRpseeError::Custom("Internal error during `piece` call".to_string())
-            });
+    fn piece(&self, requested_piece_index: PieceIndex) -> RpcResult<Option<Vec<u8>>> {
+        let Some(archived_segment) = self.piece_cache.lock().as_ref().and_then(Weak::upgrade)
+        else {
+            return Ok(None);
+        };
 
-            if let Err(err) = &result {
-                error!(?err, %piece_index, "Failed to get a piece.");
+        let indices = archived_segment
+            .segment_header
+            .segment_index()
+            .segment_piece_indexes();
+        let pieces = &archived_segment.pieces;
+        for (piece_index, piece) in indices.into_iter().zip(pieces.iter()) {
```

Comment:
Following the question about the iterator: we iterate the array here right after filling it.